File: backend/app/schemas/sources.py
"""Data source and synchronization API contracts."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class DataSourceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    key: str
    name: str
    source_type: str
    base_url: str
    enabled: bool
    health_status: str
    sync_interval_minutes: int
    last_sync_at: datetime | None
    last_success_at: datetime | None
    last_error: str | None


class SyncRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    source_id: str
    trigger_type: str
    status: str
    started_at: datetime | None
    finished_at: datetime | None
    fetched_count: int
    created_count: int
    updated_count: int
    skipped_count: int
    error_count: int
    error_summary: str | None
    diagnostic_id: str | None
    created_at: datetime

File: frontend/src/api/sources.ts
import request from './request'

export interface DataSource {
  id: string; key: string; name: string; source_type: 'official_api' | 'rss'; base_url: string
  enabled: boolean; health_status: 'unknown' | 'healthy' | 'error'; sync_interval_minutes: number
  last_sync_at: string | null; last_success_at: string | null; last_error: string | null
}

export interface SyncRun {
  id: string; source_id: string; trigger_type: string; status: 'running' | 'succeeded' | 'failed'
  started_at: string | null; finished_at: string | null; fetched_count: number; created_count: number
  updated_count: number; skipped_count: number; error_count: number; error_summary: string | null
  diagnostic_id: string | null; created_at: string
}

export interface ConnectionResult { status: 'healthy' | 'error'; item_count: number; message: string; checked_at: string }

export const sourceApi = {
  list: async () => (await request.get<DataSource[]>('/sources')).data,
  update: async (id: string, data: { enabled?: boolean; sync_interval_minutes?: number }) => (await request.patch<DataSource>(`/sources/${id}`, data)).data,
  test: async (id: string) => (await request.post<ConnectionResult>(`/sources/${id}/test`)).data,
  sync: async (id: string) => (await request.post<SyncRun>(`/sources/${id}/sync`)).data,
  runs: async (id: string) => (await request.get<SyncRun[]>(`/sources/${id}/sync-runs`)).data,
  validateRss: async (feed_url: string) => (await request.post<ConnectionResult>('/sources/rss/validate', { feed_url })).data,
  addRss: async (data: { name: string; feed_url: string; sync_interval_minutes: number }) => (await request.post<DataSource>('/sources/rss', data)).data,
  remove: async (id: string) => request.delete(`/sources/${id}`),
}
